internal/database/postgres.go

```diff
@@ -309,66 +311,125 @@ FROM TempDownsampled;
 }
 
 // TODO: consider returning workstationGroup
-func (conn PostgresConn) LatestData() ([]uplink.GpuStatsUpload, error) {
-	// we pull Uuid twice so we can put one into Stat sample and the other into Info
-	rows, err := conn.db.Query(`SELECT g.Machine, g.Uuid, g.Uuid, g.Name,
-			g.Brand, g.DriverVersion, g.MemoryTotal,
-			s.MemoryUtilisation, s.GpuUtilisation, s.MemoryUsed,
-			s.FanSpeed, s.Temp, s.MemoryTemp, s.GraphicsVoltage,
-			s.PowerDraw, s.GraphicsClock, s.MaxGraphicsClock,
-			s.MemoryClock, s.MaxMemoryClock
+func (conn PostgresConn) LatestData() (broadcast.Workstations, error) {
+	// pull all the machines in, then gpus for those machines
+	// has to all be done in a transaction to avoid race conditions
+	tx, err := conn.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make(map[string][]broadcast.Workstation)
+	machines, err := tx.Query(`SELECT GroupName, Hostname, CPU, Motherboard,
+		Notes, LastSeen
+		FROM Machines`)
+	if err != nil {
+		return nil, errors.Join(err, tx.Rollback())
+	}
+
+	for machines.Next() {
+		var groupName *string
+		var machine broadcast.Workstation
+		var lastSeen time.Time
+
+		err = machines.Scan(&groupName, &machine.Name, &machine.CPU,
+			&machine.Motherboard, &machine.Notes, &lastSeen)
+		if err != nil {
+			return nil, errors.Join(err, tx.Rollback())
+		}
+
+		// coalesce null and empty group names to default
+		if groupName == nil || strings.TrimSpace(*groupName) == "" {
+			fallback := DefaultGroup
+			groupName = &fallback
+		}
+
+		machine.LastSeen = time.Since(lastSeen)
+		machine.Gpus = nil
+
+		groups[*groupName] = append(groups[*groupName], machine)
+	}
+
+	// check for error whilst iterating, continuing if it's "no results"
+	err = machines.Err()
+	if err != nil && !errors.Is(sql.ErrNoRows, err) {
+		return nil, errors.Join(err, tx.Rollback())
```

Comment:
NIT: Not ideal, pollutes the error returned if rollback also fails. Go wants simple error messages, and simple flow. I'd suggest logging the rollback error and returning `err` so that you can actually test the type/presence/etc of this error properly without a rollback failure getting in the way.

Reply:
I disagree that this pollutes the error. `errors.Join()` returns a wrapped error. If you're testing for the presence/type of an error, you should be using `errors.Is()`, it checks all the errors wrapped and returns true if any of the errors matched the target error.